File: sdks/js/test/live.test.ts
import { describe, expect, it } from "vitest";
import { GigasttClient } from "../src/index.js";
import type { Transcript } from "../src/index.js";

// Live-server integration test. Skipped unless GIGASTT_TEST_WS_URL points at
// a running gigastt server (which requires the model, ~850 MB):
//
//   cargo run --release -- serve   # in the repository root
//   GIGASTT_TEST_WS_URL=ws://127.0.0.1:9876/v1/ws npx vitest run test/live.test.ts

const LIVE_URL = process.env.GIGASTT_TEST_WS_URL;

/** Generates `seconds` of a 440 Hz sine tone as 16 kHz PCM16 frames (20 ms). */
function* sineFrames(seconds: number): Generator<Uint8Array> {
  const rate = 16_000;
  const frameSamples = rate / 50;
  const totalFrames = seconds * 50;
  for (let i = 0; i < totalFrames; i++) {
    const frame = new Int16Array(frameSamples);
    for (let j = 0; j < frameSamples; j++) {
      frame[j] = Math.round(
        12_000 * Math.sin((2 * Math.PI * 440 * (i * frameSamples + j)) / rate),
      );
    }
    yield new Uint8Array(frame.buffer);
  }
}

describe("GigasttClient against a live server", () => {
  it.skipIf(LIVE_URL === undefined)(
    "completes a ready -> audio -> stop -> final session",
    async () => {
      const finals: Transcript[] = [];
      const errors: string[] = [];

      let resolveFinal!: () => void;
      const finalPromise = new Promise<void>((resolve) => {
        resolveFinal = resolve;
      });

      const client = await GigasttClient.connect(LIVE_URL as string, {
        sampleRate: 16_000,
        handlers: {
          onFinal: (t) => {
            finals.push(t);
            resolveFinal();
          },
          onError: (e) => errors.push(`${e.code}: ${e.message}`),
        },
      });

      expect(client.ready.model).not.toBe("");

      for (const frame of sineFrames(2)) {
        client.sendPCM(frame);
        await new Promise((resolve) => setTimeout(resolve, 10));
      }
      client.stop();

      await Promise.race([
        finalPromise,
        new Promise((_, reject) =>
          setTimeout(() => reject(new Error("no final within 30s")), 30_000),
        ),
      ]);

      expect(errors).toEqual([]);
      expect(finals.length).toBeGreaterThanOrEqual(1);
      expect(finals.at(-1)?.is_final).toBe(true);

      client.close();
    },
    45_000,
  );
});

File: sdks/go/live_test.go
package gigastt

import (
	"context"
	"encoding/binary"
	"math"
	"os"
	"testing"
	"time"
)

// TestLiveServer exercises the full wire protocol against a live gigastt
// server. It is skipped unless GIGASTT_TEST_WS_URL points at a running server
// (which requires the model, ~850 MB):
//
//	cargo run --release -- serve   # in the repository root
//	GIGASTT_TEST_WS_URL=ws://127.0.0.1:9876/v1/ws go test -run Live -v
func TestLiveServer(t *testing.T) {
	url := os.Getenv("GIGASTT_TEST_WS_URL")
	if url == "" {
		t.Skip("GIGASTT_TEST_WS_URL not set; skipping live-server integration test")
	}

	finals := make(chan Transcript, 1)
	c, err := Dial(context.Background(), url,
		WithSampleRate(16000),
		WithHandlers(Handlers{
			OnFinal: func(tr Transcript) { finals <- tr },
			OnError: func(se *ServerError) { t.Errorf("server error: %v", se) },
		}),
	)
	if err != nil {
		t.Fatalf("Dial: %v", err)
	}
	defer c.Close()

	if c.Ready().Model == "" {
		t.Error("ready.model is empty")
	}

	// Stream 2 seconds of a 440 Hz sine tone as 16 kHz PCM16 in 20ms frames.
	const (
		rate       = 16000
		frameSamps = rate / 50
		duration   = 2 * time.Second
	)
	frame := make([]byte, frameSamps*2)
	totalFrames := int(duration.Seconds() * 50)
	for i := 0; i < totalFrames; i++ {
		for j := 0; j < frameSamps; j++ {
			sample := int16(12000 * math.Sin(2*math.Pi*440*float64(i*frameSamps+j)/rate))
			binary.LittleEndian.PutUint16(frame[j*2:], uint16(sample))
		}
		if err := c.SendPCM(frame); err != nil {
			t.Fatalf("SendPCM: %v", err)
		}
		time.Sleep(10 * time.Millisecond)
	}

	if err := c.Stop(); err != nil {
		t.Fatalf("Stop: %v", err)
	}

	select {
	case f := <-finals:
		if !f.IsFinal {
			t.Error("final transcript has is_final=false")
		}
		t.Logf("live final transcript: %q (%d words)", f.Text, len(f.Words))
	case <-time.After(30 * time.Second):
		t.Fatal("no final transcript within 30s")
	}
}
